lucene/core/src/java/org/apache/lucene/util/graph/GraphTokenStreamFiniteStrings.java

```diff
@@ -0,0 +1,294 @@
+ * the graph token stream is converted to an {@link Automaton} and from there we use a {@link FiniteStringsIterator} to collect the various
+ * token streams for each finite string.
+ */
+public class GraphTokenStreamFiniteStrings {
+  /* TODO:
+     Most of this is a combination of code from TermAutomatonQuery and TokenStreamToTermAutomatonQuery. Would be
+     good to make this so it could be shared. */
+  private final Automaton.Builder builder;
+  Automaton det;
+  private final Map<BytesRef, Integer> termToID = new HashMap<>();
+  private final Map<Integer, BytesRef> idToTerm = new HashMap<>();
+  private int anyTermID = -1;
+
+  public GraphTokenStreamFiniteStrings() {
+    this.builder = new Automaton.Builder();
+  }
+
+  private static class BytesRefArrayTokenStream extends TokenStream {
+    private final BytesTermAttribute termAtt = addAttribute(BytesTermAttribute.class);
+    private final BytesRef[] terms;
+    private int offset;
+
+    BytesRefArrayTokenStream(BytesRef[] terms) {
+      this.terms = terms;
+      offset = 0;
+    }
+
+    @Override
+    public boolean incrementToken() throws IOException {
+      if (offset < terms.length) {
+        clearAttributes();
+        termAtt.setBytesRef(terms[offset]);
+        offset = offset + 1;
+        return true;
+      }
+
+      return false;
+    }
+  }
+
+  /**
+   * Gets the list of finite string token streams from the given input graph token stream.
+   */
+  public List<TokenStream> getTokenStreams(final TokenStream in) throws IOException {
+    // build automation
+    build(in);
+
+    List<TokenStream> tokenStreams = new ArrayList<>();
+    final FiniteStringsIterator finiteStrings = new FiniteStringsIterator(det);
+    for (IntsRef string; (string = finiteStrings.next()) != null; ) {
+      final BytesRef[] tokens = new BytesRef[string.length];
+      for (int idx = string.offset, len = string.offset + string.length; idx < len; idx++) {
+        tokens[idx - string.offset] = idToTerm.get(string.ints[idx]);
+      }
+
+      tokenStreams.add(new BytesRefArrayTokenStream(tokens));
+    }
+
+    return tokenStreams;
+  }
+
+  private void build(final TokenStream in) throws IOException {
+    if (det != null) {
+      throw new IllegalStateException("Automation already built");
+    }
+
+    final TermToBytesRefAttribute termBytesAtt = in.addAttribute(TermToBytesRefAttribute.class);
+    final PositionIncrementAttribute posIncAtt = in.addAttribute(PositionIncrementAttribute.class);
+    final PositionLengthAttribute posLengthAtt = in.addAttribute(PositionLengthAttribute.class);
+    final OffsetAttribute offsetAtt = in.addAttribute(OffsetAttribute.class);
+
+    in.reset();
+
+    int pos = -1;
+    int lastPos = 0;
+    int maxOffset = 0;
+    int maxPos = -1;
+    int state = -1;
+    while (in.incrementToken()) {
+      int posInc = posIncAtt.getPositionIncrement();
+      assert pos > -1 || posInc > 0;
+
+      if (posInc > 1) {
```

Comment:
@dsmiley That was actually pulled out of the existing [TokenStreamToTermAutomatonQuery.java](https://github.com/apache/lucene-solr/blob/master/lucene/sandbox/src/java/org/apache/lucene/search/TokenStreamToTermAutomatonQuery.java#L77).  Let me look into it more.